Grpc handlers

Lighty0410 opened the pull request:
Description:

This PR fixes/closes the following issues:
Closes:#33

Before submitting this PR I have done the following:
- [x] wrote all possible unit tests
- [x] checked linter locally
- [x] ran unit tests locally
- [x] updated swagger/readme accordingly

Lighty0410 committed:
Create handlers

Lighty0410 committed:
Add gRPC handlers. Add graceful gRPC shutdown in main.

Lighty0410 committed:
Update Makefile. 

Now it's possible to use "make proto" to create server.pb.go

Lighty0410 committed:
Create a new port for k8s.

Lighty0410 committed:
Update dockerfile. Add an additional expose for gRPC

Lighty0410 committed:
Update comenataries

Lighty0410 committed:
Update proto file, update fields name

Lighty0410 committed:
Update fields in server.proto and ekadashi.go

Lighty0410 committed:
Create handlers 


Add gRPC handlers. Add graceful gRPC shutdown in main. 


Update Makefile. 

Now it's possible to use "make proto" to create server.pb.go 
Create a new port for k8s.


Update dockerfile. Add an additional expose for gRPC


Update comenataries 


Update proto file, update fields name  


Update fields in server.proto and ekadashi.go

Lighty0410 committed:
Merge remote-tracking branch 'origin/grpc-handlers' into grpc-handlers

sashayakovtseva commented on Makefile:
```diff
@@ -1,5 +1,7 @@
 BIN_DIR := ./bin
 BUILDNAME := $(BIN_DIR)/server
+PROTO_DIR := ./pkg/server/grpc/api
+PROTOFILE := ./pkg/server/grpc/api/server.proto
```
Can be changed to `$(PROTO_DIR)/server.proto`

sashayakovtseva commented on pkg/server/grpc/api/server.proto:
```diff
@@ -0,0 +1,27 @@
+
+
+message Session {
+    string name = 1;
```
name is username? if so, change field name

sashayakovtseva commented on pkg/server/grpc/api/server.proto:
```diff
@@ -0,0 +1,27 @@
+}
+
+message User {
+    string user = 1;
```
user is name??

sashayakovtseva commented on pkg/server/grpc/api/server.proto:
```diff
@@ -0,0 +1,27 @@
+message User {
+    string user = 1;
+    string password = 2;
+    Session auth = 3;
```
Should not be here

sashayakovtseva replied:
You better create LoginRequest that will contain both user and session
And create RegisterRequest to contain only user

sashayakovtseva commented on pkg/server/grpc/api/server.proto:
```diff
@@ -0,0 +1,27 @@
+    Session auth = 3;
+}
+
+message Response {
```
Also create a separate response for each call, do not mix everything together!

sashayakovtseva commented on pkg/server/grpc/ekadashi.go:
```diff
@@ -0,0 +1,22 @@
+// ShowEkadashi send an endpoint request to the controller.
+// If succeed returns ekadashi date.
+func (s *Service) ShowEkadashi(ctx context.Context, u *api.User) (*api.Response, error) {
+	if u.Auth.Name != "session_token" {
```
No need for this, only valid for cookies.

sashayakovtseva commented on pkg/server/grpc/register.go:
```diff
@@ -0,0 +1,55 @@
+		return nil, fmt.Errorf("gRPC request failed: %v", err)
+	}
+	return &api.Response{
+		Response: "register succeed!",
```
No need for this response here, can be empty

sashayakovtseva commented on pkg/server/grpc/register.go:
```diff
@@ -0,0 +1,55 @@
+	if err != nil {
+		return nil, err
+	}
+	return &api.Response{
```
Same here

sashayakovtseva commented on pkg/server/grpc/server.go:
```diff
@@ -0,0 +1,28 @@
+		log.Println("cannot connect to gRPC: ", err)
+	}
+	service := NewService(c)
+	implementation := api.EkadashiServer(service)
```
what is this??

sashayakovtseva commented on pkg/server/grpc/server.go:
```diff
@@ -0,0 +1,28 @@
+	service := NewService(c)
+	implementation := api.EkadashiServer(service)
+	server := grpc.NewServer()
+	api.RegisterEkadashiServer(server, implementation)
```
no need to convert to interface

sashayakovtseva commented on pkg/server/grpc/server.go:
```diff
@@ -0,0 +1,28 @@
+	implementation := api.EkadashiServer(service)
+	server := grpc.NewServer()
+	api.RegisterEkadashiServer(server, implementation)
+	err = server.Serve(listener)
```
This is a blocking call!

Lighty0410 committed:
Refactor code

Lighty0410 committed:
Replace fields

sashayakovtseva commented on cmd/app/main.go:
```diff
@@ -36,10 +36,10 @@ func main() {
-	log.Printf("Shutting down due to signal: %v", sig)
 	err = httpServer.Shutdown(context.Background())
+	grpcServer.GracefulStop()
+	log.Printf("Shutting down due to signal: %v", sig)
```
Why you moved it here? It was fine to log right after signal is received

sashayakovtseva commented on pkg/server/grpc/ekadashi.go:
```diff
@@ -9,14 +9,14 @@ import (
-		return nil, fmt.Errorf("incorrect session name")
+func (s *Service) ShowEkadashi(ctx context.Context, u *api.Session) (*api.ShowEkadashiResponse, error) {
+	if u.Token == "" {
+		return nil, fmt.Errorf("cannot handle an empty token")
```
cannot handle an empty token -> auth token is required

sashayakovtseva commented on pkg/server/grpc/helper.go:
```diff
@@ -12,13 +12,13 @@ var validUsername = regexp.MustCompile(`^[a-zA-Z0-9]+$`).MatchString
 		return fmt.Errorf("field username contain latin characters and numbers without space only")
 	}
-	if !validPassword(u.Password) {
+	if !validPassword(u.Name) {
```
validPassword(u.Name)???
PAY ATTENTION to what you do!

sashayakovtseva commented on pkg/server/grpc/helper.go:
```diff
@@ -12,13 +12,13 @@ var validUsername = regexp.MustCompile(`^[a-zA-Z0-9]+$`).MatchString
 		return fmt.Errorf("field password contain latin characters and numbers without space only")
 	}
-	if len(u.User) < minSymbols {
+	if len(u.Password) < minSymbols {
```
SAME here!

sashayakovtseva commented on pkg/server/grpc/register.go:
```diff
@@ -23,33 +23,29 @@ func NewService(c *controller.Controller) *Service {
-	err = s.controller.RegisterUser(controller.User{Username: u.User, Password: u.Password})
+	err = s.controller.RegisterUser(controller.User{Username: u.User.Name, Password: u.User.Password})
 	if err != nil {
 		return nil, fmt.Errorf("gRPC request failed: %v", err)
```
could not register user

sashayakovtseva commented on pkg/server/grpc/register.go:
```diff
@@ -23,33 +23,29 @@ func NewService(c *controller.Controller) *Service {
-	return &api.Response{
-		Response: "register succeed!",
-	}, nil
+	return nil, nil
```
nil, nil will panic!
Did you test it ??

sashayakovtseva commented on pkg/server/grpc/register.go:
```diff
@@ -23,33 +23,29 @@ func NewService(c *controller.Controller) *Service {
-	session, err := s.controller.LoginUser(controller.User{Username: u.User, Password: u.Password})
+	session, err := s.controller.LoginUser(controller.User{Username: u.User.Name, Password: u.User.Password})
 	if err != nil {
 		return nil, err
```
WRAP error!

sashayakovtseva commented on pkg/server/grpc/server.go:
```diff
@@ -17,12 +16,13 @@ func NewGrpcServer(c *controller.Controller) (*grpc.Server, error) {
+	go func() {
+		err = server.Serve(listener)
+		if err != nil {
+			log.Printf("cannot to listen to gRPC server: %v", err)
```
cannot is used without TO

sashayakovtseva commented on pkg/server/grpc/ekadashi.go:
```diff
@@ -9,14 +9,14 @@ import (
-	date, err := s.controller.ShowEkadashi(u.Auth.Token)
+	date, err := s.controller.ShowEkadashi(u.Token)
 	if err != nil {
 		return nil, err
```
WRAP error

sashayakovtseva commented on pkg/server/grpc/server.go:
```diff
@@ -0,0 +1,28 @@
+
+// NewGrpcServer sets up a new TCP route and creates a new gRPC server.
+func NewGrpcServer(c *controller.Controller) (*grpc.Server, error) {
+	listener, err := net.Listen("tcp", ":50051")
```
Better accept port in params

Lighty0410 committed:
Refactor code. 

Add more descriptive errors, refactor main, update proto- file, update ekadashi

Lighty0410 committed:
Add address  param for NewGrpcServer

Lighty0410 committed:
Update protofile and response for better view

Lighty0410 committed:
Refactor code, update proto file
